redis/src/cluster_async/mod.rs

```diff
@@ -867,17 +868,38 @@ where
             .await
             .map(|(result, _)| result),
             Some(ResponsePolicy::OneSucceededNonEmpty) => {
-                future::select_ok(receivers.into_iter().map(|(_, receiver)| {
-                    Box::pin(async move {
-                        let result = convert_result(receiver.await)?;
-                        match result {
-                            Value::Nil => Err((ErrorKind::ResponseError, "no value found").into()),
-                            _ => Ok(result),
-                        }
-                    })
-                }))
-                .await
-                .map(|(result, _)| result)
+                // Attempt to return the first result that isn't `Nil` or an error.
+                // If no such response is found and all servers returned `Nil`, it indicates that all shards are empty, so return `Nil`.
+                // If we received only errors, return the last received error.
+                // If we received a mix of errors and `Nil`s, we can't determine if all shards are empty,
```

Comment:
this decision is tightly couples `OneSucceededNonEmpty` with the semantics of the RANDOMKEY command, where returning Nil implies that all DB nodes are empty. If there'll be another command that uses `OneSucceededNonEmpty`, then this decision might be incorrect in such a case.

I don't know if there ever will be another command using `OneSucceededNonEmpty`, but the fact that this decision is essentially for RANDOMKEY should be clearly explained in the documentation.

Reply:
It might be as simple as renaming this to: `FirstSucceededNonEmpty`
RANDOMKEY really doesn't care which node succeeded, but the first encountered is enough to satisfy the command.

Reply:
renamed to FirstSucceededNonNilOrAllNil